src/storage/src/source/generator.rs

```diff
@@ -111,42 +110,50 @@ impl SourceRender for LoadGeneratorSourceConnection {
                     .map(MzOffset::decode_row),
             );
 
-            let Some(mut offset) = resume_upper.into_option() else {
+            let Some(resume_offset) = resume_upper.into_option() else {
                 return;
             };
-            cap.downgrade(&offset);
 
-            let mut rows = as_generator(&self.load_generator, self.tick_micros)
-                .by_seed(mz_ore::now::SYSTEM_TIME.clone(), None);
-
-            // Skip forward to the requested offset.
-            let tx_count = usize::cast_from(offset.offset);
-            let txns = rows
-                .by_ref()
-                .filter(|(_, typ, _, _)| matches!(typ, GeneratorMessageType::Finalized))
-                .take(tx_count)
-                .count();
-            assert_eq!(txns, tx_count, "produced wrong number of transactions");
+            let mut rows = as_generator(&self.load_generator, self.tick_micros).by_seed(
+                mz_ore::now::SYSTEM_TIME.clone(),
+                None,
+                resume_offset,
+            );
 
             let tick = Duration::from_micros(self.tick_micros.unwrap_or(1_000_000));
 
-            while let Some((output, typ, value, diff)) = rows.next() {
-                let message = (
-                    output,
-                    Ok(SourceMessage {
-                        upstream_time_millis: None,
-                        key: (),
-                        value,
-                        headers: None,
-                    }),
-                );
-
-                data_output.give(&cap, (message, offset, diff)).await;
-
-                if matches!(typ, GeneratorMessageType::Finalized) {
-                    offset += 1;
-                    cap.downgrade(&offset);
-                    tokio::time::sleep(tick).await;
+            while let Some((output, event)) = rows.next() {
+                match event {
+                    Event::Message(offset, (value, diff)) => {
+                        let message = (
+                            output,
+                            Ok(SourceMessage {
+                                upstream_time_millis: None,
+                                key: (),
+                                value,
+                                headers: None,
+                            }),
+                        );
+
+                        // Some generators always reproduce their TVC from the beginning which can
+                        // generate a significant amount of data that will overwhelm the dataflow.
+                        // Since those are not required downstream we eagerly ignore them here.
+                        if resume_offset <= offset {
```

Comment:
this seems reasonable to me, but just in case, shouldn't we periodically yield? this could be going through millions of `next` calls

(of course, when we change things to streams, itll all work better)

Reply:
We should, yeah, but this is as good as the code on main and I don't want to spend more time on generators so if it's ok I would punt on this for some future time